Implement variable check for some boundary restricted objects

lindsayad opened the pull request:
Makes sure that variable dependencies have degrees of freedom on
all nodes/element-sides along the boundaries of a boundary restricted object

Closes #9734

lindsayad committed:
Implement variable check for some boundary restricted objects

Makes sure that variable dependencies have degrees of freedom on
all nodes along the boundaries of a boundary restricted object

Closes #9734

lindsayad commented:
@tophmatthews

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1048097/) on 20700d0 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/21045/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1048378/) on 699682f wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>dfb020</th>
      <th colspan=4>#21045 699682</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>82.57%</td>
      <td>82.60%</td>
      <td>+0.03%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>75219</td>
      <td>75302</td>
      <td>+83</td>
      <td>84</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15878</td>
      <td>15864</td>
      <td>-14</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/21045/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/21045/coverage/framework)


## Modules coverage

### Tensor mechanics

<table>
  <thead>
    <tr>
      <th></th>
      <th>dfb020</th>
      <th colspan=4>#21045 699682</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>89.48%</td>
      <td>89.48%</td>
      <td>-0.00%</td>
      <td>-</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>26215</td>
      <td>26214</td>
      <td>-1</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>3081</td>
      <td>3081</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/21045/coverage_diff/tensor_mechanics)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/21045/coverage/tensor_mechanics)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/21045/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/21045/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/21045/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/21045/coverage/contact)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/21045/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/21045/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/21045/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/21045/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/21045/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/21045/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/21045/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/21045/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/21045/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/21045/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/21045/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/21045/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/21045/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/21045/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/21045/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/21045/coverage/richards)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/21045/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/21045/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/21045/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/21045/coverage/xfem)

</details>



This comment will be updated on new commits.

GiudGiud reviewed (commented):
should we have a timed section for this? Or we expect negligible?

I think we should have a parameter to skip it too

GiudGiud reviewed (commented):
mean to start discussions, not additional work if unreasonable

GiudGiud commented on test/tests/misc/boundary_variable_check/tests:
```diff
@@ -0,0 +1,20 @@
+      type = RunException
+      input = test.i
+      expect_err = "'bad' of type 'SelfAux' depends on variable 'v'. However, that variable does not appear to have a superset of that object's block/boundary restriction."
+      detail = 'auxiliary kernels'
```
I guess not done in the checks

lindsayad replied:
I could add nodal bcs. I would have have to create a `BoundarySideIntegrityCheckThread` in order to correctly validate integrated bcs and interface kernels. And maybe I should :frowning_face:

GiudGiud commented:
Can you please add documentation to this page?
https://mooseframework.inl.gov/framework_development/sanity_checking.html#!

lindsayad commented on test/tests/misc/boundary_variable_check/tests:
```diff
@@ -0,0 +1,20 @@
+  design = 'AuxKernels/index.md UserObjects/index.md'
+  [error]
+    requirement = 'The system shall error if a boundary restricted object depends on a variable that is not defined on any portion of the boundary. These boundary restricted objects include'
+    [aux]
+      type = RunException
+      input = test.i
+      expect_err = "'bad' of type 'SelfAux' depends on variable 'v'. However, that variable does not appear to have a superset of that object's block/boundary restriction."
+      detail = 'auxiliary kernels'
+    []
+    [uo]
+      type = RunException
+      input = test.i
+      expect_err = "'bad' of type 'NodalExtremeValue' depends on variable 'v'. However, that variable does not appear to have a superset of that object's block/boundary restriction."
+      cli_args = "AuxKernels/active='' Postprocessors/active='bad'"
+      detail = 'nodal user objects'
+    []
```
what new requirement/detail would that add? Actually that would lead to no errors because this integrity check is for nodeset based objects and because the variables are all well defined on the nodeset along the interface, there would be no errors

GiudGiud replied:
well you would have to restrict the variables as well.
I m thinking of distinguishing between internal sideset checks and external (to the mesh) checks in the test

lindsayad replied:
If you only have two domains, there's no way to restrict a Lagrange variable such that it's not defined on the interface nodeset

GiudGiud replied:
yeah then you d need 3 domains

lindsayad replied:
So you want me to add a test for what kind of object(s) on an internal boundary?

lindsayad commented on framework/src/variables/MooseVariableDependencyInterface.C:
```diff
@@ -0,0 +1,37 @@
+}
+
+void
+MooseVariableDependencyInterface::checkEvaluable(const libMesh::DofObject & dof_object) const
```
I'll go with `checkVariables`

lindsayad committed:
Add boundary name to error message

Also change `checkEvaluable` to `checkVariables`

lindsayad committed:
Add checking of nodal boundary conditions

lindsayad committed:
Put all check code into the loop

Allows the DRY rule

lindsayad committed:
Add BoundaryElemIntegrityCheckThread

Test
- integrated bcs
- boundary elemental aux kernels
- side user objects

lindsayad committed:
Thread safety for BoundaryNodeIntegrityCheckThread

lindsayad commented:
This has turned into a disaster. I'm not sure these checks are worth the effort

lindsayad commented:
users do all sorts of crazy stuff in their objects which can't really blame them for. Looks like there is crazy stuff done in both `SideUserObjects` and `IntegratedBCs`. So I'm inclined to just not check those

lindsayad committed:
Omit nodal bc variable from coverage check

We do not execute nodal bcs if there are no dof indices for the
nodal bc variable on a node, so it's safe to omit the variable
from checking

lindsayad committed:
Default nodal object checking to true and side object checking to false

lindsayad commented:
@hugary1995 will need to use the `boundary_restricted_node_integrity_check = false` flag for his Raccoon tutorial that is failing here. The right boundary node that the integrity check is complaining about is not initially in the domain of `disp_x` but I'm guessing that after execution of `CoupledVarThresholdElementSubdomainModifier`, then the node *is* in the domain of `disp_x` ... ?

lindsayad committed:
Documentation

GiudGiud commented on framework/src/bcs/ADNodalBC.C:
```diff
@@ -87,12 +89,15 @@ template <typename T>
 void
 ADNodalBCTempl<T>::computeJacobian()
 {
+  const std::vector<dof_id_type> & cached_rows = _var.dofIndices();
+  if (cached_rows.empty())
+    return;
```
why do we need this now?

lindsayad replied:
It was potential landmine waiting to happen. We return early from residual evaluation if there are no dof indices, but we were not for the Jacobian

GiudGiud commented on framework/src/bcs/ArrayNodalBC.C:
```diff
@@ -73,6 +73,9 @@ ArrayNodalBC::computeJacobian()
 void
 ArrayNodalBC::computeOffDiagJacobian(const unsigned int jvar_num)
 {
+  if (!_var.isNodalDefined())
+    return;
```
same question.
And why only the off diagonal ?

GiudGiud commented on modules/doc/content/framework_development/sanity_checking.md:
```diff
@@ -33,6 +33,18 @@ MOOSE performs several sanity checks just before the simulation runs. This inclu
 - All pluggable systems must refer to the right type of variable (scalar, vector, array, "nonlinear", "auxiliary", etc)
 - Material properties must be defined on every block or boundary where they are consumed.
 - Multiple material properties with the same name be defined on overlapping blocks/boundaries.
+- Variable definition on boundary restricted objects. Nodal object integrity
+  checking is controlled with
```
integrity or coverage?
we call it coverage for variables and materials right

lindsayad replied:
that is the parameter name for those. But there is also `fv_bcs_integrity_check`. And then all these coverage/integrity checks occur through the `CheckIntegrityAction` and `FEProblemBase::checkProblemIntegrity`

GiudGiud replied:
kk I dont mind either way. This parameter is not to be used by many users anyway

GiudGiud commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -180,6 +182,16 @@ FEProblemBase::validParams()
+      "Set to false to disable checking of boundary restricted nodal object variable dependencies, "
+      "e.g. are the variable dependencies defined on the selected boundaries?");
+  params.addParam<bool>("boundary_restricted_elem_integrity_check",
+                        false,
```
why false?
Too expensive?

lindsayad replied:
causes too many errors because there are side user objects and integrated boundary conditions (one example is this view-factor thingy in the head conduction module) that do wild things with non-local variable evaluations. But I was thinking I might introduce a `virtual` method on `BoundaryRestrictable` called `bool shouldCheckVariableDependencies() const` that some of these wild objects can override to return `false`. It kind of seems silly to have a sanity check that doesn't run by default ... almost no users would ever know to turn it on

GiudGiud reviewed (commented):
test I was thinking of was for internal sidesets.
so a mesh like:
block 1 | block 2 | block 3
block 1 has the variables / restrictions
boundary used in the objects is between block 2 and 3. 

another one is partial boundary coverage.
so in your test the 'top' boundary extends outside of the domain of definition of u/v

again just ideas

hugary1995 commented:
@lindsayad So we are settled with the parameter name `boundary_restricted_node_integrity_check` right? If so, I'll patch RACCOON right now so that it'll be forward compatible.

lindsayad commented:
I like that parameter name. I think you'll have to temporarily put `allow_unused` in your `tests` spec until this moose update percolates through into your app

moosebuild commented:
Job [External app tests](https://civet.inl.gov/job/1051107/) on c9de04d : invalidated by @hugary1995

I just updated raccoon/master, let me make sure I did the right thing.

lindsayad committed:
Allow objects to turn off integrity check

This concurrently allows us to run the boundary element check by
default. I think that we should not have sanity checks off by
default

GiudGiud commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -180,6 +182,16 @@ FEProblemBase::validParams()
+      "e.g. are the variable dependencies defined on the selected boundaries?");
+  params.addParam<bool>("boundary_restricted_elem_integrity_check",
+                        true,
+                        "Set to true to enable checking of boundary restricted elemental object "
```
```suggestion
                        "Set to false to disable checking of boundary restricted elemental object "
```

GiudGiud commented on modules/doc/content/framework_development/sanity_checking.md:
```diff
@@ -33,6 +33,18 @@ MOOSE performs several sanity checks just before the simulation runs. This inclu
+  checking is controlled with
+  [!param](/Problem/FEProblem/boundary_restricted_node_integrity_check). Elemental object
+  integrity checking is controlled with
+  [!param](/Problem/FEProblem/boundary_restricted_elem_integrity_check). The former
+  parameter defaults to `true` while the latter defaults to `false`. Checked objects include
```
```suggestion
  [!param](/Problem/FEProblem/boundary_restricted_elem_integrity_check). Checked objects include
```

GiudGiud reviewed (commented):
glad these checks pass for most objects!

lindsayad committed:
Fix documentation, given default integrity check change

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

lindsayad committed:
Internal boundary test

lindsayad committed:
Partial boundary coverage test

lindsayad commented:
Ok I think I've hit the tests you wanted

lindsayad commented:
@GiudGiud good thing you requested those tests! 😄

lindsayad committed:
Nodal bc checking must be run un-threaded :-(

lindsayad commented:
That's what we had before and that can fail in threaded runs because thread 0 may not have any nodes that are uncovered, so you can get a false negative

lindsayad commented:
I'd much rather add more code and have it function correctly/independently of the number of threads

GiudGiud commented:
yeah functioning correctly is definitely the goal here. 
sounds good with this change then.

lindsayad commented:
@GiudGiud any other concerns here?

lindsayad commented:
Thanks for your reviews @GiudGiud !
